Update Discover content for 2.0 features (#8177)
opensearch-project/documentation-website: _dashboards/discover/exploring-query-enhancements.md

> ---

Author: @sejli Please tech review this documentation. It's the new content. All other files are changes to navigation and don't require tech approval. Thank you.

layout: default
title: Exploring query enhancements
parent: Analyzing data
nav_order: 20
---

# Exploring query enhancements

> Introduced 2.17

Author: ```suggestion
Introduced 2.18
```

{: .label .label-purple }

Starting with OpenSearch 2.17, query enhancements have been made. These enhancements are experimental and may be subject to change or instability. Enhancements include the following: 

- Query languages PPL and SQL, with **Query Assist** for PPL
- Multiline query editor for PPL and SQL and autocomplete for PPL and DQL
- Data selector with **Data Explorer** that supports index patterns, indexes, and Amazon S3 connections and data configuration that supports selecting the appropriate query language based on the data type
- Link sharing through URLs without needing write permission to create an index pattern

> All existing functionality remains in OpenSearch 2.17, and the new features are designed to improve your data exploration experience. The **Enable query enhancements** is currently not on a minimum distribution. <SME: Do we need to add text about plugins?>. You can leave your feedback at <insert-link> to help the OpenSearch Project improve.

Author: @sejli Will you provide me with the link for users to leave feedback about these features? Also, please review this paragraph. Do we need to add text about plugins?

Author: @sejli Confirming this is on for 2.18. Do we have updates to the draft text?

Author: @sejli No one has provided any updates. The developers write the initial drafts for new features and the tech writer refines the content. Any updates developers can make in the PR.

> All existing functionality remains in OpenSearch 2.17, and the new features are designed to improve your data exploration experience. The **Enable query enhancements** is currently not on a minimum distribution. <SME: Do we need to add text about plugins?>. You can leave your feedback at <insert-link> to help the OpenSearch Project improve.

Reviewer: - In a distribution of just `OpenSearch` and `OpenSearch Dashboards`, query enhancements will not provide access to PPL or SQL. This requires the OpenSearch SQL plugin. I think we can make it known that query enhancements does work with just OpenSearch, but will require the SQL plugin and others to make SQL, PPL, and external datasource queries work.
- Other than making a note for getting SQL and PPL to work, looks good to me. May need some more input from @kavilla as well.

Author: Revised the Note text to add this information.

> All existing functionality remains in OpenSearch 2.17, and the new features are designed to improve your data exploration experience. The **Enable query enhancements** is currently not on a minimal distribution. In a minimal distribution of OpenSearch and OpenSearch Dashboards, query enhancements do not provide access to PPL or SQL functionality. To use these features, the [OpenSearch SQL plugin]({{site.url}}{{site.baseurl}}/search-plugins/sql/settings/) is required. Query enhancements do work with a basic OpenSearch installation, but to enable SQL, PPL, and external datasource queries, additional plugins such as the SQL plugin are necessary. Ensure that you have the [required plugins]({{site.url}}{{site.baseurl}}/install-and-configure/plugins/) installed to take full advantage of these query enhancement features. You can leave your feedback at [https://forum.opensearch.org/)](https://forum.opensearch.org/) to help the OpenSearch Software Foundation improve this feature.

Author: ```suggestion
All existing functionality remains in OpenSearch 2.17, and the new features are designed to improve your data exploration experience. The **Enable query enhancements** is currently not on a minimal distribution. In a minimal distribution of OpenSearch and OpenSearch Dashboards, query enhancements do not provide access to PPL or SQL functionality. To use these features, the [OpenSearch SQL plugin]({{site.url}}{{site.baseurl}}/search-plugins/sql/settings/) is required. Query enhancements do work with a basic OpenSearch installation, but to enable SQL, PPL, and external datasource queries, additional plugins such as the SQL plugin are necessary. Ensure that you have the [required plugins]({{site.url}}{{site.baseurl}}/install-and-configure/plugins/) installed to take full advantage of these query enhancement features. You can leave your feedback at [https://forum.opensearch.org/)](https://forum.opensearch.org/) to help the OpenSearch open source project improve this feature.
```

> All existing functionality remains in OpenSearch 2.17, and the new features are designed to improve your data exploration experience. The **Enable query enhancements** is currently not on a minimal distribution. In a minimal distribution of OpenSearch and OpenSearch Dashboards, query enhancements do not provide access to PPL or SQL functionality. To use these features, the [OpenSearch SQL plugin]({{site.url}}{{site.baseurl}}/search-plugins/sql/settings/) is required. Query enhancements do work with a basic OpenSearch installation, but to enable SQL, PPL, and external datasource queries, additional plugins such as the SQL plugin are necessary. Ensure that you have the [required plugins]({{site.url}}{{site.baseurl}}/install-and-configure/plugins/) installed to take full advantage of these query enhancement features. You can leave your feedback at [https://forum.opensearch.org/)](https://forum.opensearch.org/) to help the OpenSearch open source project improve this feature.

Author: ```suggestion
All existing functionality remains in OpenSearch 2.17, and the new features are designed to improve your data exploration experience. The **Enable query enhancements** is currently not on a minimal distribution. In a minimal distribution of OpenSearch and OpenSearch Dashboards, query enhancements do not provide access to PPL or SQL functionality. To use these features, the [OpenSearch SQL plugin]({{site.url}}{{site.baseurl}}/search-plugins/sql/settings/) is required. 

Query enhancements work with a basic OpenSearch installation, but to enable SQL, PPL, and external data source queries, additional plugins such as the SQL plugin are necessary. Ensure that you have the [required plugins]({{site.url}}{{site.baseurl}}/install-and-configure/plugins/) installed to take full advantage of these query enhancement features. 

You can leave your feedback at [https://forum.opensearch.org/)](https://forum.opensearch.org/) to help the OpenSearch open source project improve this feature.
```

---
layout: default
title: Exploring query enhancements
parent: Analyzing data
nav_order: 20
---

# Exploring query enhancements
Introduced 2.18
{: .label .label-purple }

Starting with OpenSearch 2.17, query enhancements have been made. These enhancements are experimental and may be subject to change or instability. Enhancements include the following: 

- Query languages PPL and SQL, with **Query Assist** for PPL

> - Multiline query editor for PPL and SQL and autocomplete for PPL and DQL

Author: ```suggestion
- Multiline query editor for PPL and SQL and autocomplete for PPL and DQL
- Expand/collapse query editor for multiline/single-line mode
```

> - Multiline query editor for PPL and SQL and autocomplete for PPL and DQL

Author: @kavilla Please see revised text. You're welcome to make any necessary changes to the text. Thank you!

> - Data selector with **Data Explorer** that supports index patterns, indexes, and Amazon S3 connections and data configuration that supports selecting the appropriate query language based on the data type

Reviewer 2: is it there enough to call out that index patterns aren't required for exploring your data within your indexes within OpenSearch Dashboards but still required if you want to create visualizations

Author: Does this communicate the appropriate message: `OpenSearch Dashboards allows for exploring your data within your indexes without using index patterns`

> - Link sharing through URLs without needing write permission to create an index pattern

Reviewer 2: creating an index pattern with write permission is required which was a blocker for exploring data. non-privileged users would have to depend on someone with permission to create the index pattern for the data to be explored leading to suboptimal index patterns to be created as it would be easier to just do something like `logs-*` instead of the specific logs. so i think it is right but there's a couple number of features that get unlocked without the required write permissions

Author: @kavilla @sejli Should the link sharing bullet be modified to read: Link sharing through URLs without needing write permission to create an index pattern, enabling non-privileged users to access and interact with the data, even if they don't have the ability to define the underlying index patterns

Reviewer: LGTM, @kavilla?

> All existing functionality remains in OpenSearch 2.17, and the new features are designed to improve your data exploration experience. The **Enable query enhancements** is currently not on a minimal distribution. In a minimal distribution of OpenSearch and OpenSearch Dashboards, query enhancements do not provide access to PPL or SQL functionality. To use these features, the [OpenSearch SQL plugin]({{site.url}}{{site.baseurl}}/search-plugins/sql/settings/) is required. 

Reviewer 2: > The **Enable query enhancements** is currently not on a minimal distribution.

it should be. but there is a few bugs

> Query enhancements work with a basic OpenSearch installation, but to enable SQL, PPL, and external data source queries, additional plugins such as the SQL plugin are necessary. Ensure that you have the [required plugins]({{site.url}}{{site.baseurl}}/install-and-configure/plugins/) installed to take full advantage of these query enhancement features. 

Reviewer 2: Perhaps I'm being overly cautious since you did mention it. But did we want to add a sentence that really highlights that it is the external data source that needs the SQL plugin installed. So each one of the data sources need to have the SQL plugin installed. not just the default cluster that they run their usual queries.

The following tutorials guide you through some of the experimental features and capabilities.

### Query language enhancements

You can now use PPL in **Discover**. Follow these steps to try out the feature:

1. Go to **Discover** and select **PPL** from the query language dropdown menu in the upper-right search bar. You should see a dashboard containing the query editor, histogram, and data table panes.
2. Select a sample dataset. For this example, select `opensearch_dashboards_sample_data_ecommerce` from the data source dropdown menu above the query editor and adjust the time filter to **Last 1 year**.
3. Enter the following example PPL query:

```json
source = opensearch_dashboards_sample_data_logs
| where tags = "success" 
| where geo.dest = "US"
```
{% include copy-curl.html %}

4. View the resulting output that shows the number of successful log entries originating from the United States. You should see an updated histogram and data table following the query editor. 
5. Select the **Recent queries** option within the query editor toolbar to display your recent queries.

PPL and DQL provide an autocomplete option that suggests field names, functions, and syntax. 

> ## Data sources and data types enhancements

Author: ```suggestion
## Selecting data sources and data types through the UI
```

You can now select your data sources and types from within the **Discover** dashboard. Follow these steps to try out the feature.

1. From the **Discover** page, select a data source from the dropdown menu in the upper toolbar. 

> 2. Select the **View all available data** button to display a list of your available data sources. You may need to refresh your page to display any newly added data sources.

Author: @kavilla I'm open to suggestions. Please make a suggested rewrite and I'll accept changes. Thank you.

3. Select the desired data source and follow steps displayed in the data sources window to manage your data source.

> You can now use **Query Assist** with PPL queries. With **Query Assist**, you can ask questions like _Are there any errors in my logs?_. The assistant includes predefined prompts. Follow these steps to try out the feature:

Reviewer 2: didn't see it but should we call out that Query Assist for PPL but not necessarily available for all data sources

Author: @kavilla Should we tell user which specific data sources/types it's available for?

Reviewer: Query Assist is only available in query enhancements when the data source itself supports it. It's dependent on if the user adds a cluster with query assist enabled/set up. I think we can call out that query assist is required on the OpenSearch data source, something like query assist is available when included data source has it available?

---
layout: default
title: Exploring query enhancements
parent: Analyzing data
nav_order: 20
---

# Exploring query enhancements
Introduced 2.18
{: .label .label-purple }

Starting with OpenSearch 2.17, query enhancements have been made. These enhancements are experimental and may be subject to change or instability. Enhancements include the following: 

- Query languages PPL and SQL, with **Query Assist** for PPL
- Multiline query editor for PPL and SQL and autocomplete for PPL and DQL
- Expand/collapse query editor for multiline/single-line mode

> - Data selector with **Data Explorer** that supports index patterns, indexes, and Amazon S3 connections and data configuration that supports selecting the appropriate query language based on the data type

Author: ```suggestion
- Data selector with **Data Explorer** that supports index patterns, indexes, and Amazon S3 connections and data configuration that supports selecting the appropriate query language based on the data type
- OpenSearch Dashboards allows for exploring your data within your indexes without using index patterns
```

---
layout: default
title: Exploring query enhancements
parent: Analyzing data
nav_order: 20
---

# Exploring query enhancements
Introduced 2.18
{: .label .label-purple }

Starting with OpenSearch 2.17, query enhancements have been made. These enhancements are experimental and may be subject to change or instability. Enhancements include the following: 

- Query languages PPL and SQL, with **Query Assist** for PPL
- Multiline query editor for PPL and SQL and autocomplete for PPL and DQL

> - Expand/collapse query editor for multiline/single-line mode

Author: ```suggestion
- Query editor expand/collapse for multiline/single-line mode
```

> - Data selector with **Data Explorer** that supports index patterns, indexes, and Amazon S3 connections and data configuration that supports selecting the appropriate query language based on the data type

Author: ```suggestion
- **Data Explorer** feature that supports diverse data sources, such as index patterns, indexes, and Amazon S3 connections, with built-in support for selecting the appropriate query language for your chosen data type
```

- OpenSearch Dashboards allows for exploring your data within your indexes without using index patterns
- Link sharing through URLs without needing write permission to create an index pattern

> All existing functionality remains in OpenSearch 2.17, and the new features are designed to improve your data exploration experience. The **Enable query enhancements** is currently not on a minimal distribution. In a minimal distribution of OpenSearch and OpenSearch Dashboards, query enhancements do not provide access to PPL or SQL functionality. To use these features, the [OpenSearch SQL plugin]({{site.url}}{{site.baseurl}}/search-plugins/sql/settings/) is required. 

Author: ```suggestion
OpenSearch 2.18 builds upon existing features with new query enhancements designed to improve data exploration. However, these enhancements, including PPL and SQL functionality, are not available in minimal distributions and require the separate installation of the [OpenSearch SQL plugin]({{site.url}}{{site.baseurl}}/search-plugins/sql/settings/).
{: .note} 
```

> Query enhancements work with a basic OpenSearch installation, but to enable SQL, PPL, and external data source queries, additional plugins such as the SQL plugin are necessary. Ensure that you have the [required plugins]({{site.url}}{{site.baseurl}}/install-and-configure/plugins/) installed to take full advantage of these query enhancement features. 

Author: ```suggestion
Query enhancements work with a basic OpenSearch installation, but to enable SQL, PPL, and external data source queries, additional plugins such as the SQL plugin are necessary. Index patterns require write permissions, but OpenSearch Dashboards enables data exploration through shared links, even for non-privileged users, though the SQL plugin must be installed on each external data source. Ensure that you have the [required plugins]({{site.url}}{{site.baseurl}}/install-and-configure/plugins/) installed to take full advantage of these query enhancement features. 
```

```
./bin/opensearch-dashboards --uiSettings.overrides['query:enhancements:enabled']=true
```
{% include copy-curl.html %}

## Using the experimental features

The following tutorials guide you through some of the experimental features and capabilities.

### Query language enhancements

You can now use PPL in **Discover**. Follow these steps to try out the feature:

1. Go to **Discover** and select **PPL** from the query language dropdown menu in the upper-right search bar. You should see a dashboard containing the query editor, histogram, and data table panes.
2. Select a sample dataset. For this example, select `opensearch_dashboards_sample_data_ecommerce` from the data source dropdown menu above the query editor and adjust the time filter to **Last 1 year**.
3. Enter the following example PPL query:

```json
source = opensearch_dashboards_sample_data_logs
| where tags = "success" 
| where geo.dest = "US"
```
{% include copy-curl.html %}

4. View the resulting output that shows the number of successful log entries originating from the United States. You should see an updated histogram and data table following the query editor. 

> 5. Select the **Recent queries** option within the query editor toolbar to display your recent queries.

Author: ```suggestion
5. Select the **Recent queries** option within the query editor toolbar to display your recent queries. This is available for non-DQL and Lucene queries.
```

Introduced 2.18
{: .label .label-purple }

Starting with OpenSearch 2.17, query enhancements have been made. These enhancements are experimental and may be subject to change or instability. Enhancements include the following: 

- Query languages PPL and SQL, with **Query Assist** for PPL
- Multiline query editor for PPL and SQL and autocomplete for PPL and DQL
- Query editor expand/collapse for multiline/single-line mode
- **Data Explorer** feature that supports diverse data sources, such as index patterns, indexes, and Amazon S3 connections, with built-in support for selecting the appropriate query language for your chosen data type
- OpenSearch Dashboards allows for exploring your data within your indexes without using index patterns
- Link sharing through URLs without needing write permission to create an index pattern

OpenSearch 2.18 builds upon existing features with new query enhancements designed to improve data exploration. However, these enhancements, including PPL and SQL functionality, are not available in minimal distributions and require the separate installation of the [OpenSearch SQL plugin]({{site.url}}{{site.baseurl}}/search-plugins/sql/settings/).
{: .note} 

> Query enhancements work with a basic OpenSearch installation, but to enable SQL, PPL, and external data source queries, additional plugins such as the SQL plugin are necessary. Ensure that you have the [required plugins]({{site.url}}{{site.baseurl}}/install-and-configure/plugins/) installed to take full advantage of these query enhancement features. 

Author: ```suggestion
While query enhancements work with a standard OpenSearch installation, enabling SQL, PPL, and external data source queries requires additional plugins, particularly the SQL plugin. For optimal use of these query enhancement features, make sure to install the [required plugins]({{site.url}}{{site.baseurl}}/install-and-configure/plugins/) across all relevant data sources. 
```
